### DISCUSSION TITLE
VM preparation

### DISCUSSION THREAD

**Discussion by hadiidbouk** [codemagic-ci-cd/codemagic-docs#1593] [Q&A] [answered]
Hello,

I am wondering how the process of preparing a VM works, everything is wiped off after the build is finished?

Is there any place that keeps my changes? Let’s say I added a file to the home directory `~/.my-file.txt`, this file can still be kept after the build is finished?

How does Codemagic allocate the VMs? Can other clients use the same VM in the future?

**Comment by dtrdic** [comment, marked as answer]
Hello there!

So at the beginning of each build, your source code is cloned to a new virtual machine instance where the build is run. It is destroyed immediately after the build has completed, leaving no trace of your source code or secret variables.So, no, other clients can't use the same VM in the future.

Any added files or variables are not persisted.

To access API keys, certificates, tokens, and other sensitive values during your build, an unlimited number of encrypted secrets can easily be added to the Codemagic dashboard. Only the build history log and build artifacts will be available in the Codemagic dashboard after the build finishes.

And if you want to speed up your builds you can store your dependencies on Codemagic, you can check examples on this page https://docs.codemagic.io/knowledge-codemagic/caching/
